sdks/java/core/src/test/java/org/apache/beam/sdk/values/RowTest.java

```diff
@@ -477,6 +477,172 @@ public void testCreateMapWithRowValue() {
     assertEquals(data, row.getMap("map"));
   }
 
+  @Test
```

Comment:
Is it worth having a logical type to test as well?

Reply:
yes will add one

Reply:
What get's people confused on the current API is that you can't loop over the values of a Row with `getValue()` and can't supply them to the Builder as the `addValue()` expects the input type, while the getValue returns a base type.

Reply:
added test

Reply:
still struggling to figure out a good answer to the API. getValues is supposed to return the base value so that transforms that don't know about the specific logical value can process it.

Reply:
What about this,
- `addValue`, `getValue` and `getValues` return the input type(s) (so the logical type)
- `add`, `get`, `detach` returns the base types (is also inline with the `attach`)
That means a behaviour change only for `getValue` and `getValues`

Reply:
Elaborating on the above comment:

`getValues` is currently not safe (it makes the Row mutable), if the meaning changes that it returns the logical type the implementation will need to make a copy of the content as it needs to translate the basetype to the logicaltype. As it's a copy the Row is immutable again. See this for test:

```  
  @Test
  public void testImmutability() {
    Schema schema = Schema.builder().addInt32Field("a").addInt32Field("b").build();
    Row row1 = Row.withSchema(schema).addValue(1).addValue(2).build();
    List<Object> values = row1.getValues();
    values.set(1, 3);
    assertEquals(2,(int)row1.getValue(1));
  }
```
Then the question is left for `attach` or `detach`: if you make values an immutable array it can be returned with `detach`, as it's immutable that's safe. The `attach` needs then make a copy of the array to an immutable array (or check if it's immutable) and `build` needs to make the immutable array as well.

-

Reply:
The fact that getValues returns the underlying objects I think is in line with many other things. e.g. ImmutableList in Java still returns the underlying objects, so you can mutate them. The onus is on the user not to do that.

In particular, the elements of a PCollection are not allowed to be mutated in a ParDo (and the main point of the Row object is to be used in ParDos). The DirectRunner has a mutability checker that will fail your test if you mutate the input elements.

Reply:
Aside from the mutability/. WDYT of the proposed `addValue`, `getValue` and `getValues` vs `add`, `get`, `detach`

Reply:
I propose to take this discussion out of this PR, as I really like this already. For me it's already a LGTM. (Best to wait till the 2.20 is cut though).